src/Cryptol/Parser.y

```diff
@@ -302,11 +302,22 @@ mbDoc                   :: { Maybe (Located Text) }
   : doc                    { Just $1 }
   | {- empty -}            { Nothing }
 
+propguards_cases :: { [([Prop PName], Expr PName)] }
+  : propguards_case propguards_cases { $1 ++ $2 }
+  | propguards_case { $1 }
+
+propguards_case :: { [([Prop PName], Expr PName)] }
+  : '|' propguards_quals '=>' expr { [($2, $4)] }
+
+propguards_quals :: { [Prop PName] }
+  : type {% fmap thing (mkProp $1) }
 
 decl                    :: { Decl PName }
   : vars_comma ':' schema  { at (head $1,$3) $ DSignature (reverse $1) $3   }
   | ipat '=' expr          { at ($1,$3) $ DPatBind $1 $3                    }
   | '(' op ')' '=' expr    { at ($1,$5) $ DPatBind (PVar $2) $5             }
+  | var apats_indices propguards_cases
+                           { mkIndexedPropGuardsDecl $1 $2 $3 }
   | var apats_indices '=' expr
```

Comment:
It might be a bit of a corner case, but it probably makes sense to allow guarded constants.  For example:
```

isZero : {n}  Bit
isZero
  | n == 0 => True
  | ()     => False
```

Reply:
that's quite an interesting idea

Reply:
However, it seems that there is an error that will arise from ambiguous type variables. For example:
```
isFin : {n} Bit
isFin | fin n => True
      | () => False
```
yields the error
```
[error] at tests/constraint-guards/constant.cry:2:1--2:6:
  Ambiguous numeric type: type argument 'n' of 'Main::isFin(fin n)'
```
Is that something you'd want to have check for an appearance of the type variable in the constraint guards before raising?

Reply:
Actually, you can see from the name of the function that raises the error `sFin(fin n)`, that its one of the declarations generated by expanding the prop guards. It doesn't know locally, currently, that itself came from a prop guard expansion

Reply:
actually, this is fixed by doing [your other request](https://github.com/GaloisInc/cryptol/pull/1380#discussion_r958026018). Now it just gives warnings about unused type variables